### DISCUSSION TITLE
Script to pack multiple images into one texture channel?

### DISCUSSION THREAD

**Discussion by ItsCubeTime** [ImageMagick/ImageMagick#3330] [Help]
Hi, so I have this simple formula

15*16+13=253 (storing the values 15 and 13)
253%16=13 (unpacking 13)
(253-(253%16))/16=15 (unpacking 15)

which lets me pack multiple values into one.

And I basically wanna use this to store multiple 8 bit BW textures and multiple channels of RGB textures down to one channel in a texture.

So if I have imageA and imageB (BW). How would I be able to use the above formula to pack these 2 into one channel into a separate output image?

Looking for cmd line solutions if possible (alternatively Python, tho not prefered). Im on Windows

**Comment by snibgo** [comment]
Your formulae show an example of numbers to base 16.

I'm not sure what you mean by storing "multiple channels of RGB textures down to one channel in a texture."

If you have a colour image, with RGB channels, you can separate the channels and append them, and store that in a single greayscale image. Is that what you mean?

Perhaps you can show examples of what you want.

**Reply by ItsCubeTime** [reply]
That append thing sounds about right, but how does it turn store those channels into one (and how would I extract it later on)?

The images are pbr textures for use inside of a game engine, so I'm planning on extracting the data (restoring it to the original format) at runtime in-engine (which I could then do using some kind of math expression).

Basically the goal here is to be able to store 14 isch pbr/texture channels inside of a single high but depth texture (and also lower the bit depth of the stored data along the way).

**Comment by snibgo** [comment]
Separate and append:
```
magick toes.png -channel RGB -separate +channel +append +repage x.png
```
![x](https://user-images.githubusercontent.com/33812368/109563725-1348e100-7ad8-11eb-8ea2-af363b29d3ce.png)

Chop the result into three images and combine them into RGB channels:
```
magick x.png -crop 3x1@ +repage -combine x2.png
```
![x2](https://user-images.githubusercontent.com/33812368/109563766-1fcd3980-7ad8-11eb-8d51-38faa69c0369.png)

But why are you doing this? This does not reduce file sizes.

> Basically the goal here is to be able to store 14 isch pbr/texture channels inside of a single high but depth texture (and also lower the bit depth of the stored data along the way).

Sorry, I don't understand that.

**Reply by ItsCubeTime** [reply]
Hi and thanks for the demonstration (but I dont think append was quite what I was looking for).

Like Idk how close you looked at the 3 math expressions I put in my original post. Basically 15, 13, 253, 16 could be replaced with a variable each, x, y, z, q.

Then these rules would apply:
  Condition: 0 <= x <= q-1 and 0 <= y <= q-1 and x = floor(x) and y = floor(y)

  x*q+y=z (storing the values x and y)
  z%q=y (unpacking y)
  (z-(z%q))/q=x (unpacking x)

And to fullfill those conditions for a 8-bit depth image´(x and y represents our input BW images in this case) all you would have to do is multiply x and y respectively with q (q being the bit depth you wanna store, but in base 10 instead of binary). And then in order to actually store the output data you would then of course have to divide z with z^2.

You could try visualizing the way Im storing the values as a 2D grid if you like, x representing rows and y representing columns (or the other way around).

So basically Im trying to store multiple pixels in one pixel of the image channel.

But if you dont quite understand what it does tho thats okay. the important part of my question is just how I could use a math expression (and use multiple input images inside of that expression) (like the one I use for packing in my example) to store that it one image.

I already found my way to: https://imagemagick.org/script/fx.php which seems to be what Im looking for. But there are no examples there showing how I could use multiple input images in my expression, which is where the hog for me is.

> But why are you doing this? This does not reduce file sizes.

You might be right, one could argue this is just an experiment (and saving space in VRAM is one of the goals of this, Im hoping that the compression algorithms used in my game engine will compress one image at a higher bit-depth better than several images at a lower bit-depth and that way save memory space(Since the game engine uses lossy compression Im quite sure I will save space, even if it might just be a little)) but another benefit is that it usually takes longer to load 2 smaller images than it does to load 1 larger image at runtime in a game-engine. And when you wanna load houndreds of relatively large images stuff like this can affect the user experience. And lastly it benefits the artists workflow (well my workflow as Im not collaborating with anyone rn) slightly as there will be less images (less files) that has to be moved between softwares and it would also result in a slightly less cluttered file hierarchy.

**Comment by snibgo** [comment]
Yes, two numbers that each need N bits can be packed into a number that takes 2xN bits. So two same-sized images that need 8 bits/channel/pixel could be packed into one image of the same size with 16 bits/channel/pixel.

Lossy compression relies on low-order bits being relatively unimportant, so these are liable to be changed. This would damage the image stored in those bits, so lossy compression of packed images is probably a bad idea.

For using multiple images in a "-fx" expression, see https://imagemagick.org/script/fx.php , the section headed "Source Images". If you have just two input images, I suggest using "u" and "v".

"-fx" is slow. For performance, it is better to use "-evaluate", "-compose Mathematics", and so on.

A complication is that IM internally promotes pixel values. For example, if your IM is Q16 non-HDRI, it stores pixel values as 16-bit integers, in the range [0,65535]. When it reads 8-bit files range [0,255] it promotes values to 16 bits by replicating the byte, which is multiplying by 65535/255, which is 257, which is hex #0101.

**Reply by ItsCubeTime** [reply]
Right, smart as I am I didn't even think about that my technique would break with lossy compression. Thanks for pointing this out, I've made sure to disable it the game engine for now.

> A complication is that IM internally promotes pixel values. For example, if your IM is Q16 non-HDRI, it stores pixel values as 16-bit integers, in the range [0,65535]. When it reads 8-bit files range [0,255] it promotes values to 16 bits by replicating the byte, which is multiplying by 65535/255, which is 257, which is hex #101.

Im a little confused about this, where does it store values as 16-bit integers? The version of image magick that I downloaded was the portable Q16 HDRI (with HDRI).

I did a test using -fx:

``magick albedo.jpg normal.jpg -colorspace RGB -depth 16 -fx "((u*15*16)+v*15)/255" fx_navy.exr``

As you can see its following my first formula, but Im not quite managing to extract the data correctly. The only difference to what Im doing here is supposedly that I multiply the input images with 15 to get it from 0-1 range to 0-15 range (including the min and max values in my ranges).

Can you spot what Im doing wrong here? 

I should mention that albedo.jpg is stored in sRGB and normal.jpg is stored in linear color space. fx_navy is supposed to be linear and in my game engine Ive setup the following "material" as its called to extract the values (my game engine is setup to read the texture as linear already):

![image](https://user-images.githubusercontent.com/20190653/109663936-ef83aa80-7b6c-11eb-9608-3d691f66e674.png)



In this screenshot Im trying to extract my normal.jpg (which is a image that is mainly blue) from the fx_navy.exr (node to the very left). Its basically a recreation of z%q=y (Fmod in the image is modulo) but also multiplying and dividing by 255 to get the ranges right. The image at the bottom is the original normal.jpg (the same that I fed into my IM -fx cmd), however the output Im seeing (image to the right) isnt close to matching.


Ive looked over my code several times and to me it looks good (other than the fact that Im trying to read my albedo, which is sRGB as linear (but that shouldnt stop me from extracting my normal.jpg still))

What Im using is Unreal Engines procedural/node based shader creation system, which basically lets you do different kind of image and rendering effects.

**Comment by snibgo** [comment]
> The version of image magick that I downloaded was the portable Q16 HDRI (with HDRI).

This means IM will record pixel values in memory as floating-point, in the nominal range 0.0 to 65535.0. For most purposes, we don't care what the maximum value actually is, we just refer to "QuantumRange". We often refer to a percentage of that value, eg 0% or 100%. In "-fx", IM normalises values to 1.0, so 100% of QuantumRange becomes 1.0. The result of the "-fx" expression should usually be on the range 0.0 to 1.0.

Please link to albedo.jpg and normal.jpg.

**Reply by ItsCubeTime** [reply]
Here you go.

![normal](https://user-images.githubusercontent.com/20190653/109672745-9e2be900-7b75-11eb-97cd-5a683271d463.jpg)
![Albedo](https://user-images.githubusercontent.com/20190653/109672740-9d935280-7b75-11eb-89bd-03ed84b53e96.jpg)
